File: tools/Python/t4-Python-api/src/t4login/client/chart_client.py
"""HTTP client for the T4 Chart API (barchart + tradehistory)."""

from __future__ import annotations

from io import BytesIO
from typing import Any

import httpx

from t4login.definitions.chartdata.chart_data_stream_reader import ChartDataStreamReader
from t4login.definitions.chartdata.chart_data_stream_reader_aggr import (
    ChartDataHandler,
    ChartDataStreamReaderAggr,
)
from t4login.definitions.chartdata.chart_data_type import ChartDataType, Tick as _TickDataType
from t4login.datetime_.n_date_time import NDateTime

_DEFAULT_BASE_URL = "https://api-sim.t4login.com/chart"

# T4Bin/T4BinAggr SOF record signature: <length:7bit><tag=CTAG_SOF=1><version=1 as 4-byte LE int>.
# Aggregated (T4BinAggr) SOF length = 5 (tag + version).
# Non-aggregated (T4Bin) SOF length = 13 (tag + version + 8-byte datetime).
# The HTTP binary response wraps the T4Bin payload in an envelope (header + GUID +
# request metadata) followed by the embedded T4Bin blob — we locate the blob by
# scanning for either SOF signature.
_T4BINAGGR_SOF_SIGNATURE = b"\x05\x01\x01\x00\x00\x00"
_T4BIN_SOF_SIGNATURE = b"\x0d\x01\x01\x00\x00\x00"


def _extract_t4bin_payload(content: bytes) -> bytes:
    """Strip the HTTP binary envelope and return the embedded T4Bin payload.

    The HTTP response wraps the T4Bin blob in an envelope (header + GUID +
    request metadata).  We locate the embedded blob by scanning for either
    SOF signature:

    * T4BinAggr SOF (aggregated barchart):   ``\\x05\\x01\\x01\\x00\\x00\\x00``
    * T4Bin SOF     (non-aggregated history): ``\\x0d\\x01\\x01\\x00\\x00\\x00``

    Raises:
        ValueError: If the response body is non-empty but contains no
            recognisable SOF signature — this indicates a corrupt or
            unexpected response rather than a legitimate empty result.
    """
    if not content:
        return b""
    aggr_idx = content.find(_T4BINAGGR_SOF_SIGNATURE)
    bin_idx = content.find(_T4BIN_SOF_SIGNATURE)
    candidates = [i for i in (aggr_idx, bin_idx) if i >= 0]
    if not candidates:
        raise ValueError(
            f"No T4Bin SOF signature found in {len(content)}-byte response payload. "
            "The server may have returned an error body or an unrecognised format."
        )
    return content[min(candidates):]


class ChartClient:
    """Client for the T4 Chart Data API.

    Parameters
    ----------
    token : str
        Bearer token for authentication.
    base_url : str, optional
        Base URL for the chart API (defaults to the sim environment).
    http_client : httpx.Client | None, optional
        Pre-configured httpx client. If None, one is created internally.
    """

    def __init__(
        self,
        token: str,
        *,
        base_url: str = _DEFAULT_BASE_URL,
        http_client: httpx.Client | None = None,
    ) -> None:
        self._token = token
        self._base_url = base_url.rstrip("/")
        self._owns_client = http_client is None
        self._client = http_client or httpx.Client(timeout=30.0)

    def close(self) -> None:
        """Close the underlying HTTP client (only if we created it)."""
        if self._owns_client:
            self._client.close()

    def __enter__(self) -> "ChartClient":
        return self

    def __exit__(self, *_: Any) -> None:
        self.close()

    # ------------------------------------------------------------------
    # /chart/barchart
    # ------------------------------------------------------------------

    def get_barchart_json(
        self,
        *,
        exchange_id: str,
        contract_id: str,
        chart_type: str = "Bar",
        bar_interval: str = "Minute",
        bar_period: int = 1,
        trade_date_start: str,
        trade_date_end: str,
        market_id: str | None = None,
        continuation_type: str | None = None,
        reset_interval: str | None = None,
    ) -> dict[str, Any]:
        """Fetch barchart data as JSON (default response format)."""
        params = self._barchart_params(
            exchange_id=exchange_id,
            contract_id=contract_id,
            chart_type=chart_type,
            bar_interval=bar_interval,
            bar_period=bar_period,
            trade_date_start=trade_date_start,
            trade_date_end=trade_date_end,
            market_id=market_id,
            continuation_type=continuation_type,
            reset_interval=reset_interval,
        )
        resp = self._get("/barchart", params=params, accept="application/json")
        return resp.json()  # type: ignore[no-any-return]

    def get_barchart_binary(
        self,
        *,
        exchange_id: str,
        contract_id: str,
        chart_type: str = "Bar",
        bar_interval: str = "Minute",
        bar_period: int = 1,
        trade_date_start: str,
        trade_date_end: str,
        market_id: str | None = None,
        continuation_type: str | None = None,
        reset_interval: str | None = None,
        handler: ChartDataHandler,
    ) -> None:
        """Fetch barchart data in binary format and decode via ChartDataStreamReaderAggr.

        The decoded records are dispatched to the provided *handler*.
        """
        params = self._barchart_params(
            exchange_id=exchange_id,
            contract_id=contract_id,
            chart_type=chart_type,
            bar_interval=bar_interval,
            bar_period=bar_period,
            trade_date_start=trade_date_start,
            trade_date_end=trade_date_end,
            market_id=market_id,
            continuation_type=continuation_type,
            reset_interval=reset_interval,
        )
        resp = self._get("/barchart", params=params, accept="application/octet-stream")
        payload = _extract_t4bin_payload(resp.content)
        ChartDataStreamReaderAggr.read(payload, handler)

    # ------------------------------------------------------------------
    # /chart/tradehistory
    # ------------------------------------------------------------------

    def get_tradehistory_json(
        self,
        *,
        exchange_id: str,
        contract_id: str,
        market_id: str | None = None,
        trade_date_start: str | None = None,
        trade_date_end: str | None = None,
        start: str | None = None,
        end: str | None = None,
        since: str | None = None,
    ) -> dict[str, Any]:
        """Fetch trade history data as JSON (default response format).

        Date range selection — use **one** of the following:

        * ``trade_date_start`` / ``trade_date_end`` — request all ticks for
          one or more full trading sessions identified by their trade date
          (``YYYY-MM-DD`` strings).  The T4 *trade date* for a session that
          spans midnight is the date on which the session is listed, not the
          wall-clock date of each tick.

        * ``start`` / ``end`` — request ticks within an absolute intraday
          time window (ISO-8601 datetime strings).  Use this when you need a
          slice of a session rather than a whole day.

        * ``since`` — request all ticks since a given UTC datetime (ISO-8601).
          Useful for incremental polling.

        Supplying both ``trade_date_*`` and ``start``/``end`` at the same time
        is not validated here and the API will give precedence to one set;
        avoid mixing them.
        """
        params = self._tradehistory_params(
            exchange_id=exchange_id,
            contract_id=contract_id,
            market_id=market_id,
            trade_date_start=trade_date_start,
            trade_date_end=trade_date_end,
            start=start,
            end=end,
            since=since,
        )
        resp = self._get("/tradehistory", params=params, accept="application/json")
        return resp.json()  # type: ignore[no-any-return]

    def get_tradehistory_binary(
        self,
        *,
        exchange_id: str,
        contract_id: str,
        market_id: str | None = None,
        trade_date_start: str | None = None,
        trade_date_end: str | None = None,
        start: str | None = None,
        end: str | None = None,
        since: str | None = None,
        data_type: ChartDataType = _TickDataType,
    ) -> ChartDataStreamReader:
        """Fetch trade history in binary format and return a ChartDataStreamReader.

        The caller iterates the reader via ``reader.read()`` to decode individual
        records from the non-aggregated T4Bin stream.

        Date range selection — same rules as :meth:`get_tradehistory_json`:
        supply either ``trade_date_start``/``trade_date_end`` for full trading
        sessions, *or* ``start``/``end`` for an intraday slice, *or* ``since``
        for incremental polling.  Do not mix the two date styles.
        """
        params = self._tradehistory_params(
            exchange_id=exchange_id,
            contract_id=contract_id,
            market_id=market_id,
            trade_date_start=trade_date_start,
            trade_date_end=trade_date_end,
            start=start,
            end=end,
            since=since,
        )
        resp = self._get("/tradehistory", params=params, accept="application/octet-stream")
        payload = _extract_t4bin_payload(resp.content)
        stream = BytesIO(payload)
        reader = ChartDataStreamReader(
            stream=stream,
            trade_date=NDateTime(0),
            market_id=market_id or "",
            data_type=data_type,
        )
        return reader

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _get(self, path: str, *, params: dict[str, str], accept: str) -> httpx.Response:
        headers = {
            "Authorization": f"Bearer {self._token}",
            "Accept": accept,
        }
        url = f"{self._base_url}{path}"
        resp = self._client.get(url, params=params, headers=headers)
        resp.raise_for_status()
        return resp

    @staticmethod
    def _barchart_params(
        *,
        exchange_id: str,
        contract_id: str,
        chart_type: str,
        bar_interval: str,
        bar_period: int,
        trade_date_start: str,
        trade_date_end: str,
        market_id: str | None,
        continuation_type: str | None,
        reset_interval: str | None,
    ) -> dict[str, str]:
        params: dict[str, str] = {
            "exchangeId": exchange_id,
            "contractId": contract_id,
            "chartType": chart_type,
            "barInterval": bar_interval,
            "barPeriod": str(bar_period),
            "tradeDateStart": trade_date_start,
            "tradeDateEnd": trade_date_end,
        }
        if market_id is not None:
            params["marketID"] = market_id
        if continuation_type is not None:
            params["continuationType"] = continuation_type
        if reset_interval is not None:
            params["resetInterval"] = reset_interval
        return params

    @staticmethod
    def _tradehistory_params(
        *,
        exchange_id: str,
        contract_id: str,
        market_id: str | None,
        trade_date_start: str | None,
        trade_date_end: str | None,
        start: str | None,
        end: str | None,
        since: str | None,
    ) -> dict[str, str]:
        params: dict[str, str] = {
            "exchangeId": exchange_id,
            "contractId": contract_id,
        }
        if market_id is not None:
            params["marketID"] = market_id
        if trade_date_start is not None:
            params["tradeDateStart"] = trade_date_start
        if trade_date_end is not None:
            params["tradeDateEnd"] = trade_date_end
        if start is not None:
            params["start"] = start
        if end is not None:
            params["end"] = end
        if since is not None:
            params["since"] = since
        return params

File: tools/JavaScript/JSDemo/decoder/client/ChartClient.js
/**
 * Port of `t4login.client.chart_client.ChartClient`.
 *
 * HTTP client for the T4 Chart API. Uses the global `fetch` (Node 18+ /
 * any modern browser). For JSON endpoints, returns parsed objects directly;
 * for binary endpoints, the response is decoded with the chart-data stream
 * readers.
 */

import { ByteReader } from '../connection/ByteReader.js';
import { NDateTime } from '../datetime/NDateTime.js';
import { ChartDataStreamReader } from '../definitions/chartdata/ChartDataStreamReader.js';
import { ChartDataStreamReaderAggr } from '../definitions/chartdata/ChartDataStreamReaderAggr.js';
import { Tick as _TickDataType } from '../definitions/chartdata/ChartDataType.js';

const DEFAULT_BASE_URL = 'https://api-sim.t4login.com/chart';

// T4Bin / T4BinAggr SOF record signatures.
// Aggregated (T4BinAggr): length=5, tag=CTAG_SOF=1, version=1 (LE int32)
const _T4BINAGGR_SOF = Uint8Array.of(0x05, 0x01, 0x01, 0x00, 0x00, 0x00);
// Non-aggregated (T4Bin):  length=13, tag=CTAG_SOF=1, version=1
const _T4BIN_SOF      = Uint8Array.of(0x0d, 0x01, 0x01, 0x00, 0x00, 0x00);

/**
 * Locate the embedded T4Bin payload by scanning for either SOF signature.
 * Throws if no SOF is found in a non-empty response (corrupt / unexpected
 * format), mirroring the Python helper.
 *
 * @param {Uint8Array} content
 * @returns {Uint8Array}
 */
export function extractT4BinPayload(content) {
  if (content.length === 0) return content;
  const aggrIdx = _indexOf(content, _T4BINAGGR_SOF);
  const binIdx = _indexOf(content, _T4BIN_SOF);
  const candidates = [aggrIdx, binIdx].filter((i) => i >= 0);
  if (candidates.length === 0) {
    throw new Error(
      `No T4Bin SOF signature found in ${content.length}-byte response payload. ` +
      'The server may have returned an error body or an unrecognised format.',
    );
  }
  return content.subarray(Math.min(...candidates));
}

export class ChartClient {
  /**
   * @param {object} opts
   * @param {string} opts.token   Bearer token.
   * @param {string} [opts.baseUrl]
   * @param {typeof fetch} [opts.fetch]   Custom fetch (defaults to globalThis.fetch).
   */
  constructor({ token, baseUrl = DEFAULT_BASE_URL, fetch: fetchImpl } = {}) {
    if (!token) throw new Error('ChartClient: token is required');
    this._token = token;
    this._baseUrl = baseUrl.replace(/\/+$/, '');
    this._fetch = fetchImpl ?? globalThis.fetch?.bind(globalThis);
    if (typeof this._fetch !== 'function') {
      throw new Error('ChartClient: no fetch implementation available (pass `fetch` option)');
    }
  }

  // ------------------------------------------------------------------
  // /chart/barchart
  // ------------------------------------------------------------------

  async getBarchartJson(params) {
    const resp = await this._get('/barchart', this._barchartParams(params), 'application/json');
    return resp.json();
  }

  /**
   * Fetch barchart binary and dispatch decoded records to `handler`.
   * @param {object} opts  Same fields as getBarchartJson plus:
   *   - handler: chart-data handler (onBar, onMarketDefinition, ...)
   */
  async getBarchartBinary({ handler, ...rest }) {
    if (!handler) throw new Error('getBarchartBinary: handler is required');
    const resp = await this._get(
      '/barchart',
      this._barchartParams(rest),
      'application/octet-stream',
    );
    const buf = new Uint8Array(await resp.arrayBuffer());
    const payload = extractT4BinPayload(buf);
    ChartDataStreamReaderAggr.read(payload, handler);
  }

  // ------------------------------------------------------------------
  // /chart/tradehistory
  // ------------------------------------------------------------------

  async getTradehistoryJson(params) {
    const resp = await this._get(
      '/tradehistory',
      this._tradehistoryParams(params),
      'application/json',
    );
    return resp.json();
  }

  /**
   * Fetch trade history binary and return a configured ChartDataStreamReader.
   * Caller iterates `while (reader.read()) { ... }` to decode records.
   *
   * @param {object} opts
   * @returns {Promise<ChartDataStreamReader>}
   */
  async getTradehistoryBinary({ dataType = _TickDataType, ...rest }) {
    const resp = await this._get(
      '/tradehistory',
      this._tradehistoryParams(rest),
      'application/octet-stream',
    );
    const buf = new Uint8Array(await resp.arrayBuffer());
    const payload = extractT4BinPayload(buf);
    return new ChartDataStreamReader({
      data: new ByteReader(payload),
      tradeDate: new NDateTime(0n),
      marketId: rest.marketId ?? '',
      dataType,
    });
  }

  // ------------------------------------------------------------------
  // Internal
  // ------------------------------------------------------------------

  async _get(path, params, accept) {
    const url = new URL(this._baseUrl + path);
    for (const [k, v] of Object.entries(params)) {
      if (v != null) url.searchParams.set(k, String(v));
    }
    const resp = await this._fetch(url.toString(), {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${this._token}`,
        Accept: accept,
      },
    });
    if (!resp.ok) {
      const body = await resp.text().catch(() => '');
      throw new Error(`HTTP ${resp.status} ${resp.statusText}: ${body}`);
    }
    return resp;
  }

  _barchartParams({
    exchangeId,
    contractId,
    chartType = 'Bar',
    barInterval = 'Minute',
    barPeriod = 1,
    tradeDateStart,
    tradeDateEnd,
    marketId = null,
    continuationType = null,
    resetInterval = null,
  }) {
    return {
      exchangeId,
      contractId,
      chartType,
      barInterval,
      barPeriod: String(barPeriod),
      tradeDateStart,
      tradeDateEnd,
      marketID: marketId,
      continuationType,
      resetInterval,
    };
  }

  _tradehistoryParams({
    exchangeId,
    contractId,
    marketId = null,
    tradeDateStart = null,
    tradeDateEnd = null,
    start = null,
    end = null,
    since = null,
  }) {
    return {
      exchangeId,
      contractId,
      marketID: marketId,
      tradeDateStart,
      tradeDateEnd,
      start,
      end,
      since,
    };
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function _indexOf(haystack, needle) {
  if (needle.length === 0) return 0;
  outer: for (let i = 0; i <= haystack.length - needle.length; i++) {
    for (let j = 0; j < needle.length; j++) {
      if (haystack[i + j] !== needle[j]) continue outer;
    }
    return i;
  }
  return -1;
}
